### PR TITLE
StepFunctions: Multi-accounts compatibility

### PR DESCRIPTION
**Pull request by viren-nadkarni**
## Motivation

This PR introduces multi-account support to the legacy StepFunctions provider. It also makes multi-account friendly fixes to the work-in-progress StepFunctions v2 provider.

cc: @MEPalma 

## Implementation

The legacy provider uses [StepFunctions Local](https://docs.aws.amazon.com/step-functions/latest/dg/sfn-local-jar.html) which itself does not support namespacing. So we take an approach similar to Kinesis. Each account+region combo will run its own instance of SFLocal. Of course this is not optimal but should fulfill simple usecases. 

However, SFLocal does not respect the `--account-id` argument and overwrites ARNs like Lambda Functions during invocations. So I guess the way forward is to only support true multi-accounts in the new SF provider and keep this limitation in the legacy one. The multi-region parameter is maintained.

Criteria for merging: all stepfunctions tests (legacy as well as the **ci/circleci: itest-sfn-v2-provider** job)  pass when `AWS_ACCESS_KEY_ID` is set to a non-default account ID or when `AWS_REGION_NAME` is set to non-`us-east-1` region.

**79ec46e61e52940f2be5d2377b4933a77e5c2b61 must be reverted before merge**

## Related

Closes https://github.com/localstack/localstack/issues/8205

### CODE REVIEW

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_dynamodb.py [hunk 1]
```diff
@@ -124,8 +123,7 @@ def _from_error(self, env: Environment, ex: Exception) -> FailureEvent:
 
     def _eval_service_task(self, env: Environment, parameters: dict) -> None:
         api_action = camel_to_snake_case(self.resource.api_action)
-
-        dynamodb_client = connect_to(config=Config(parameter_validation=False)).dynamodb
+        dynamodb_client = get_boto_client(env, "dynamodb")
```
Comment (MEPalma) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_dynamodb.py:126]: Is this still ensuring the parameters validation is off?

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py [hunk 2]
```diff
@@ -85,10 +84,12 @@ def __init__(self, resource_arn: ResourceARN):
         self.account = resource_arn.account
 
     @staticmethod
-    def from_resource_arn(arn: str) -> Resource:
+    def from_resource_arn(account_id: str, region_name: str, arn: str) -> Resource:
```
Comment (MEPalma) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py:87]: Runtime values are ideally evaluated during execution. So resource should become an EvalComponent, evaluate its region and account in `_eval_body` using the environment, and push to the stack a resource object that includes all the values.

File: localstack/services/stepfunctions/asl/eval/environment.py [hunk 3]
```diff
@@ -26,8 +26,14 @@
 
 
 class Environment:
-    def __init__(self, context_object_init: ContextObjectInitData):
+    def __init__(
+        self, account_id: str, region_name: str, context_object_init: ContextObjectInitData
+    ):
         super(Environment, self).__init__()
+
```
Comment (MEPalma) [inline localstack/services/stepfunctions/asl/eval/environment.py:33]: I think we should decorate these in an object where the environment can expose runtime values like these 2.

File: localstack/services/stepfunctions/asl/parse/preprocessor.py [hunk 4]
```diff
@@ -191,6 +191,10 @@
 
 
 class Preprocessor(ASLParserVisitor):
+    def __init__(self, account_id: str, region_name: str):
```
Comment (MEPalma) [inline localstack/services/stepfunctions/asl/parse/preprocessor.py:194]: Adding the values to the environment is a decent solution, and evaluation components can produce runtime values like region and account during evaluation. So I don't think we should pollute the Preprocessor with runtime values, as conceptually it just builds the evaluation tree.

**Review by MEPalma — CHANGES_REQUESTED** [review]
Thank you for the changes! I would like the separation between building the evaluation tree and evaluating runtime values to remain consistent. We achieve this with Component and EvalComponent classes. Adding these runtime values to the environment is valid, but not in the preprocessor. Also a question about parameter validation of the new clients.

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_dynamodb.py (hunk 1 above)
Comment (viren-nadkarni) [reply localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_dynamodb.py:126]: Yes! Please make sure you use this helper to get the Boto client and not `connect_externally_to`

https://github.com/localstack/localstack/blob/aad9b80492fdc9f7b22f48af0184cb5febf6dd23/localstack/services/stepfunctions/backend/utils.py#L7-L13

File: localstack/services/stepfunctions/asl/parse/preprocessor.py (hunk 4 above)
Comment (viren-nadkarni) [reply localstack/services/stepfunctions/asl/parse/preprocessor.py:194]: When the `Preprocessor` constructs the `Resource` object from ARN, the ARN does not necessarily contain the region and the account ID. So it is necessary that the correct account ID and region from the request context are passed down.

https://github.com/localstack/localstack/blob/aad9b80492fdc9f7b22f48af0184cb5febf6dd23/localstack/services/stepfunctions/asl/parse/preprocessor.py#L238

Using `aws_stack.get_region()` is not an option as it doesn't always have the proper region. Moreover, I'll be deprecating that method in the future for this exact reason.

**Review by MEPalma — APPROVED** [review]
Resources are now evaluation components for `ResourceRuntimePart` which includes the region and account. This means that the request's account and region are used if none is declared in the resource Arn. All other staticResource properties are still accessible from the declaration objects themselves.

**Pull request merged**